src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -0,0 +1,561 @@
+      hiveDatabase = connectorConfig.getString(HdfsSinkConnectorConfig.HIVE_DATABASE_CONFIG);
+      this.hiveMetaStore = hiveMetaStore;
+      this.hive = hive;
+      this.executorService = executorService;
+      this.futures = futures;
+      hivePartitions = new HashSet<>();
+    }
+  }
+
+  private enum State {
+    RECOVERY_STARTED,
+    RECOVERY_PARTITION_PAUSED,
+    WAL_APPLIED,
+    WAL_TRUNCATED,
+    OFFSET_RESET,
+    WRITE_STARTED,
+    WRITE_PARTITION_PAUSED,
+    SHOULD_ROTATE,
+    TEMP_FILE_CLOSED,
+    WAL_APPENDED,
+    FILE_COMMITTED;
+
+    private static State[] vals = values();
+    public State next() {
+      return vals[(this.ordinal() + 1) % vals.length];
+    }
+  }
+
+  public boolean recover() {
+    try {
+      switch (state) {
+        case RECOVERY_STARTED:
+          pause();
+          nextState();
+        case RECOVERY_PARTITION_PAUSED:
+          applyWAL();
+          nextState();
+        case WAL_APPLIED:
+          truncateWAL();
+          nextState();
+        case WAL_TRUNCATED:
+          resetOffsets();
+          nextState();
+        case OFFSET_RESET:
+          resume();
+          nextState();
+          break;
+        default:
+          log.error("{} is not a valid state to perform recovery for topic partition {}.", state, tp);
+      }
+    } catch (ConnectException e) {
+      log.error("Recovery failed.");
+      setRetryTimeout(timeoutMs);
+      return false;
+    }
+    return true;
+  }
+
+  public void write() {
+    long now = System.currentTimeMillis();
+    if (failureTime > 0 && now - failureTime < timeoutMs) {
+      return;
+    }
+    if (state.compareTo(State.WRITE_STARTED) < 0) {
+      boolean success = recover();
+      if (!success) {
+        return;
+      }
+    }
+    while(!buffer.isEmpty()) {
+      try {
+        switch (state) {
+          case WRITE_STARTED:
+            pause();
+            nextState();
+          case WRITE_PARTITION_PAUSED:
+            if (currentSchema == null) {
+              if (compatibility != Compatibility.NONE && offset != -1) {
+                String topicDir = FileUtils.topicDirectory(url, topicsDir, tp.topic());
+                CommittedFileFilter filter = new TopicPartitionCommittedFileFilter(tp);
+                FileStatus fileStatusWithMaxOffset = FileUtils.fileStatusWithMaxOffset(storage, new Path(topicDir), filter);
+                if (fileStatusWithMaxOffset != null) {
+                  currentSchema = schemaFileReader.getSchema(conf, fileStatusWithMaxOffset.getPath());
+                }
+              }
+            }
+            SinkRecord record = buffer.peek();
+            Schema valueSchema = record.valueSchema();
+            if (shouldRotate()) {
+              nextState();
+            } else if (SchemaUtils.shouldChangeSchema(valueSchema, currentSchema, compatibility)) {
+              currentSchema = valueSchema;
+              if (hiveIntegration) {
+                if (!tableCreated) {
```

Comment:
I'm not sure this logic works. What if the table exists already? The flag `tableCreated` may not be set (this task just got assigned the topic partition but the table exists), and so we'll go into `createHiveTable()`, but I don't see a recovery path that falls back to altering the schema if the table was already created.